src/node/oscillator.rs

```diff
@@ -679,555 +325,179 @@ impl AudioProcessor for OscillatorRenderer {
-    /// * `interpol_ratios` - the interpolation ratio of each harmonics used by linear interpolatio
-    /// * `buffer` - the buffer is filled with generated wavetable data (avoid allocation)
     #[inline]
-    fn generate_wavetable(
-        norms: &[f32],
-        phases: &mut [f32],
-        incr_phases: &[f32],
-        interpol_ratios: &[f32],
-        buffer: &mut Vec<f32>,
-    ) {
-        buffer.clear();
-
-        while phases[1] <= TABLE_LENGTH_F32 {
-            let mut sample = 0.0;
-            for i in 1..phases.len() {
-                let gain = norms[i];
-                let phase = phases[i];
-                let incr_phase = incr_phases[i];
-                let mu = interpol_ratios[i];
-                // truncation is desired
-                #[allow(clippy::cast_possible_truncation)]
-                // phase + incr_phase is always positive
-                #[allow(clippy::cast_sign_loss)]
-                let idx = (phase + incr_phase) as usize;
-                let inf_idx = idx % TABLE_LENGTH_USIZE;
-                let sup_idx = (idx + 1) % TABLE_LENGTH_USIZE;
-                // Linear interpolation
-                sample += SINETABLE[inf_idx].mul_add(1. - mu, SINETABLE[sup_idx] * mu) * gain;
-                phases[i] = phase + incr_phase;
-            }
-
-            buffer.push(sample);
+    fn generate_custom(&mut self) -> f32 {
+        let periodic_wave = self.periodic_wave.as_ref().unwrap().as_slice();
+        let position = self.phase * TABLE_LENGTH_USIZE as f64;
+        let floored = position.floor();
+
+        let prev_index = floored as usize;
+        let mut next_index = prev_index + 1;
+        if next_index == TABLE_LENGTH_USIZE {
+            next_index = 0;
         }
-    }
 
-    /// normalizes the given buffer
-    fn norm_factor(buffer: &[f32]) -> f32 {
-        1. / buffer
-            .iter()
-            .copied()
-            .reduce(f32::max)
-            .expect("Maximum value not found")
+        // linear interpolation into lookup table
+        let k = (position - floored) as f32;
+        periodic_wave[prev_index].mul_add(1. - k, periodic_wave[next_index] * k)
     }
 
-    /// computes the `polyBLEP` corrections to apply to aliasing signal
-    ///
-    /// `polyBLEP` stands for `polyBandLimitedstEP`
-    fn poly_blep(&self, mut t: f32) -> f32 {
-        let dt = self.incr_phase;
-        if t < dt {
+    // computes the `polyBLEP` corrections to apply to aliasing signal
+    // `polyBLEP` stands for `polyBandLimitedstEP`
+    // This basically soften the sharp edges in square and sawtooth signals
+    // to avoid infinite frequencies impulses (jumps from -1 to 1 or inverse).
+    // cf. http://www.martin-finke.de/blog/articles/audio-plugins-018-polyblep-oscillator/
+    //
+    // @note: do not apply in tests so we can avoid relying on snapshots
+    #[inline]
+    fn poly_blep(mut t: f64, dt: f64, is_test: bool) -> f64 {
```

Comment:
This works but it is a bit uncommon I think, to pass a `test` boolean. I'll try to suggest something different later

Reply:
Yup, I agree, actually I started by just doing:

```rs
#[cfg(test)]
fn poly_blep_test() {
   if t < dt {
     t /= dt;
     t + t - t * t - 1.0
   } else if t > 1.0 - dt {
     0.0
   }
}
```

but I finally tough it would be more error prone

Reply:
Let's leave it like this then!